pybart/new_matcher.py

```diff
@@ -0,0 +1,330 @@
+#         m = matcher(doc)
+#         for conv_name in m.names():
+#             matches = m.matches_for(conv_name)
+#             transform(matches…) # TBD
+#         sentence = ...
+#
+# for doc in docs:
+#     convert_sentence(sentence, matcher)
+
+import itertools
+from collections import defaultdict
+from typing import NamedTuple, Sequence, Mapping, Any, List, Tuple
+import re
+import spacy
+from spacy.matcher import Matcher as SpacyMatcher
+from .constraints import *
+
+
+# function that checks that a sequence of Label constraints is satisfied
+# TODO - maybe move this function as a 'satisfied' method of the Label constraint? will simplify this code to
+#   "for constraint in label_constraints: if not label_constraint.satisfied(actual_labels): return False; return True"
+def are_labels_satisfied(label_constraints: Sequence[Label], actual_labels: List[str]) -> Tuple[bool, Set[str]]:
+    successfully_matched = set()
+    # we need to satisfy all constraints in the sequence, so if one fails, return False
+    for constraint in label_constraints:
+        if isinstance(constraint, HasNoLabel):
+            # check if a regex or exact match is required
+            is_regex = constraint.value.startswith('/') and constraint.value.endswith('/')
+            
+            # for each edged label, check if the label matches the constraint, and fail if it does,
+            #   because it is a negative search (that is non of the labels should match)
+            for actual_label in actual_labels:
+                if (is_regex and re.match(constraint.value[1:-1], actual_label)) or (constraint.value == actual_label):
+                    return False, set()
+        elif isinstance(constraint, HasLabelFromList):
+            current_successfully_matched = set()
+            # at least one of the constraint strings should match, so return False only if none of them did.
+            for value_option in constraint.value:
+                # check if a regex or exact match is required
+                is_regex = value_option.startswith('/') and value_option.endswith('/')
+
+                # for each edged label, check if the label matches the constraint, and store it if it does,
+                #   because it is a positive search (that is at least one label should match)
+                for actual_label in actual_labels:
+                    if (is_regex and re.match(value_option[1:-1], actual_label)) or (value_option == actual_label):
+                        # store the matched label
+                        current_successfully_matched.add(actual_label)
+            # this means no match, so fail
+            if not current_successfully_matched:
+                return False, set()
+            
+            # concat the captured labels of the current constraint with previous ones, as there can be a few positive constraints.
+            successfully_matched.update(current_successfully_matched)
+    
+    return True, successfully_matched
```

Comment:
why do you return successfully_matched? and isn't it all of them?
how will you handle it when you move to the `.satisfied(actual)` version?

Reply:
it is not all of them, as for "HasLabelFromList" it's sufficient that one label satisfies the constraint, so need to keep only those that do satisfy. And for "HasNoLabel", we actually don't need to store any label, cause we just want to validate that none of the labels matches the constraint.
so this is my suggestion (when using the new `.satisfied(actual)` method): this method will return only the set of matched labels (or an empty set for HasNoLabel) - yes, no more boolean. and when there is no match we will throw an exception.
this allows us:
1. to draw back fast (rather than returning from funcs) when there is no match. {optimization}
2. to break in the middle once a match was found on the negative lookup (without needing to finish the search) {optimization}
3. to return only one value from the `satisfied` method (instead of a tuple), which in turn allows us to still write the generic `are_labels_satisfied` (which maybe be renamed to `get_matched_labels`) function in a compactic way `[successfully_matched.update(constraint.satisfied(actual_labels)) for constraint in label_constraints]`.

but I didn't answer why do we need them - and that is a bigger question. if you recall, I wanted to have the labels that are associated with the match for the step of graph-rewriting. so this is the logical place to store them for future use. but it is really cumbersome as you can see.

Reply:
i am not too excited about using exceptions as control-flow like this, although i can see the optimization appeal (but---maybe exceptions themselves are inefficient, this needs checking). will likely be nicer to return None rather than throw an exception, but this will make your list-comprehension code more cumbersome.. so... idk.

Reply:
I consulted with the Internet and with @mcshlain and decided to let go of the exception handling. so ill resolve this discussion and add a new commit for further review